src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+    }
+
+    // variables used for tracking
+    int currPos = START_POS;
+    int numNodes = pois.size() + 1; // total nodes: #locations + center
+    LocalTime currentTime = START_TIME;
+    int count = START_COUNT;
+    Tuple cost = new Tuple(START_COST, new ArrayList<>());
+    Tuple ans = new Tuple(START_ANS, new ArrayList<>());
+    boolean[] visited = new boolean[numNodes]; // sets values to "false"  
+
+    // call function to set timeMatrix, intToPlaceId, & openHours
+    createTimeMatrix(center, pois);
+    populateIntMapAndOpenHours(center, pois);
+
+    // set central visited node to visted
+    visited[0] = true;
+
+    // call recursive helper fuction to get final Answer
+    this.finalAnswer = solverHelper(currPos, numNodes, currentTime, count, 
+        cost, ans, visited);
+  }
+
+  /**
+   * Recursive helper function that is used to solve computational part of
+   * algorithm. Called in global `solver()` method
+   *
+   * @param currPos represents the placeId of where the alg currently is
+   * @param numNodes total number of locations to visit (includes center)
+   * @param currentTime time if path at this point were taken 
+   * @param count number of pois visited during this path
+   * @param cost Tuple of current path (time cost, List of pois)
+   * @param ans the current best solution found
+   * @param array of which pois have been visited by this path
+   */
+  public Tuple solverHelper(int currPos, int numNodes, LocalTime currentTime, 
+      int count, Tuple cost, Tuple ans, boolean[] visited) {
+    /**
+     * If last node is reached and shares an edge w/ start node
+     * Calculate min of cost and currAns
+     * return and keep traversing the graph/matrix
+     */
+    if ((count == numNodes-1) && timeMatrix[currPos][0] > 0) {
+      if (ans.getCurrAns() < cost.getCurrAns() + timeMatrix[currPos][0]) {
+        return ans;
+      } else {
+        cost.incCurrAns(timeMatrix[currPos][0]);
+        return cost;
+      }
+    }
+
+    /**
+     * Backtracking Step: loop through the time Matrix from currPos, increase
+     * count by 1 and update (int) and (List) value of cost
+     */
+    for (int i = 0; i < numNodes; i++) {
+      // set open boolean and time to add to currenTime
+      boolean isOpen = (boolean) checkIsOpenSetTime(currentTime, i,
+          currPos).get(0);
+      int timeToAdd = (int) checkIsOpenSetTime(currentTime, i,
+          currPos).get(1);
+      
+      // if node is unvisited and greater than 0, i.e. not the same node
+      if (!visited[i] && timeMatrix[currPos][i] > 0 && isOpen) {
+        // mark ith node as visited
+        visited[i] = true;
+
+        // update cost
+        LocalTime timePostUpdate = currentTime.plusMinutes((long) timeToAdd);
+        List<Integer> costPath = cost.getCurrPath();
+        List<Integer> updatePath = new ArrayList<>();
+        updatePath.addAll(costPath);
+        updatePath.add(i);
```

Comment:
This is fine. Totally optional, but another way you can do this would be to add directly to cost.getCurrPath() here then remove the element you added from cost.getCurrPath below where you reset visited. This might make the code a little harder to reason about, but it reduces the storage needs of this algorithm since you'd only have 1 list instead of 1 list per recursive call.

Reply:
I had issues in the past about sending the same object through the recursive call multiple times. I think I'm going to leave as is.